src/primitives/block.cpp

```diff
@@ -109,6 +112,47 @@ uint256 CBlock::CheckMerkleBranch(uint256 hash, const std::vector<uint256>& vMer
     return hash;
 }
 
+static uint64_t next_pow2(uint64_t x)
+{
+    x -= 1;
+    x |= (x >> 1);
+    x |= (x >> 2);
+    x |= (x >> 4);
+    x |= (x >> 8);
+    x |= (x >> 16);
+    x |= (x >> 32);
+    return x + 1;
+}
+
+uint256 CBlock::BuildAuthDataMerkleTree() const
+{
+    std::vector<uint256> tree;
+    auto perfectSize = next_pow2(vtx.size());
+    tree.reserve(perfectSize * 2); // Safe upper bound for the number of total nodes.
```

Comment:
Why is this `perfectSize * 2`, rather than just `perfectSize`?

Reply:
`tree` gets its internal nodes appended to it, so for a perfect tree the bound on total nodes is 2*leaves.